src/agent/Agent.cpp

```diff
@@ -0,0 +1,222 @@
+#include "Agent.h"
+
+Agent::Agent(zsock_t* pipe) {
+    // TODO: remove connect() from here
+    redis_discover_.connect();
+
+    pipe_ = pipe;
+    request_ = NULL;
+    router_ = zsock_new(ZMQ_ROUTER);
+    poller_ = zpoller_new(pipe_);
+    zpoller_add(poller_, router_);
+    servers_ = zhash_new();
+
+    // update server list asap when Agent starts
+    next_discovery_ = zclock_mono();
+}
+
+Agent::~Agent() {
+    zsock_destroy(&router_);
+    zhash_destroy(&servers_);
+    zmsg_destroy(&request_);
+    zpoller_destroy(&poller_);
+}
+
+void Agent::unlock_servers() {
+    Server* item = (Server*) zhash_first(servers_);
+    while(item) {
+        item->busy_ = false;
+        item = (Server*) zhash_next(servers_);
+    }
+}
+
+// function to free the memory when zhash element is destroyed
+static void s_server_free (void *argument) {
```

Comment:
you don't need this function from anywhere else in the code.
so you should use a lambda function instead for zhash_freefn,
or put this one in an anonymous namespace {} so it gets hidden from other translation units (.cpp files)